src/rust-crypto/rust-crypto.ts

```diff
@@ -1280,6 +1292,52 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, CryptoEventH
         obj.signatures = Object.fromEntries(sigs.entries());
     }
 
+    /**
+     * Implementation of {@link CryptoApi#restoreKeyBackupWithPassphrase}.
+     */
+    public async restoreKeyBackupWithPassphrase(
+        phassphrase: string,
+        opts?: KeyBackupRestoreOpts,
+    ): Promise<KeyBackupRestoreResult> {
+        const backupInfo = await this.backupManager.getServerBackupInfo();
+        if (!backupInfo?.version) {
+            throw new Error("No backup info available");
+        }
+
+        const privateKey = await keyFromAuthData(backupInfo.auth_data, phassphrase);
+
+        // Cache the key
+        await this.storeSessionBackupPrivateKey(privateKey, backupInfo.version);
+        return this.restoreKeyBackup(opts);
+    }
+
+    /**
+     * Implementation of {@link CryptoApi#restoreKeyBackup}.
+     */
+    public async restoreKeyBackup(opts?: KeyBackupRestoreOpts): Promise<KeyBackupRestoreResult> {
+        const privateKeyFromCache = await this.getSessionBackupPrivateKey();
+        if (!privateKeyFromCache) throw new Error("No decryption key found in cache");
+
+        const backupInfo = await this.backupManager.getServerBackupInfo();
+        if (!backupInfo?.version) throw new Error("Missing version in backup info");
+
+        const backupDecryptor = await this.getBackupDecryptor(backupInfo, privateKeyFromCache);
+
+        opts?.progressCallback?.({
+            stage: "fetch",
+        });
+
+        let result: KeyBackupRestoreResult;
+        try {
+            result = await this.backupManager.restoreKeyBackup(backupInfo.version, backupDecryptor, opts);
+        } finally {
+            // Free to avoid to keep in memory the decryption key stored in it. To avoid to exposing it to an attacker.
+            backupDecryptor.free();
+        }
+
+        return result;
```

Comment:
Let's put the callback inside the try/finally, in case it throws.

Also, we don't need the temporary `result`.

```suggestion
        const backupDecryptor = await this.getBackupDecryptor(backupInfo, privateKeyFromCache);

        try {
            opts?.progressCallback?.({
                stage: "fetch",
            });

            return await this.backupManager.restoreKeyBackup(backupInfo.version, backupDecryptor, opts);
        } finally {
            // Free to avoid to keep in memory the decryption key stored in it. To avoid to exposing it to an attacker.
            backupDecryptor.free();
        }
```

Reply:
seriously? that works where the other doesn't? I don't understand how that is possible. Fuck, as you so succinctly put it, Javascript.

(NB it's a different case to that medium post: there is no `return` in the `finally` block. `return` in `try` blocks absolutely should be executed, whether or not there is a `finally`. They just happen *before* the finally.)

Reply:
to confirm: you did `return await this.backupManager.restoreKeyBackup`, not just `return this.backupManager.restoreKeyBackup` ?

Reply:
`return await this.backupManager.restoreKeyBackup` is working and not `return this.backupManager.restoreKeyBackup`. 
Both are returning a promise but in the first case, we are returning in the promise the result of `restoreKeyBackup` and in the second the call of `restoreKeyBackup` is *waiting* to be called.

So I suppose that in the first case, the finally block is called between the end of execution of `restoreKeyBackup` and the return.

Reply:
> `return await this.backupManager.restoreKeyBackup` is working and not `return this.backupManager.restoreKeyBackup`.
> Both are returning a promise but in the first case, we are returning in the promise the result of `restoreKeyBackup` and in the second the call of `restoreKeyBackup` is _waiting_ to be called.

yes. The important thing is to have an `await` within the `try` block.